### DISCUSSION TITLE
Direct mode, multiple include directories, and equally-named header files

### DISCUSSION THREAD

**Discussion by chelmuth** [ccache/ccache#921] [Support]
During development in a sufficiently complex project with multiple include directories (`-I`), headers files may move from one directory to another or may be shadowed by additional files with the same name and `#include_next`. In such cases *direct mode* often produces false hits and the new file is not picked up. Should direct mode support such use cases or is it common practice to configure `direct_mode = false`?

#### Example (evaluated with version 4.4)
```
> ln -s <path to>/ccache ./gcc
> mkdir inc1
> mkdir inc2
> echo '#include <file.h>' > c.c
> echo '#warning inc1' > inc1/file.h

> ./gcc -Iinc2 -Iinc1 -c -o c.o c.c
In file included from c.c:3:0:
inc1/file.h:1:2: warning: #warning inc1 [-Wcpp]
 #warning inc1
  ^~~~~~~

> echo '#warning inc2' > inc2/file.h
> ./gcc -Iinc2 -Iinc1 -c -o c.o c.c
In file included from c.c:3:0:
inc1/file.h:1:2: warning: #warning inc1 [-Wcpp]
 #warning inc1
  ^~~~~~~
> CCACHE_RECACHE=1 ./gcc -Iinc2 -Iinc1 -c -o c.o c.c
In file included from c.c:3:0:
inc2/file.h:1:2: warning: #warning inc2 [-Wcpp]
 #warning inc2
  ^~~~~~~

**Comment by jrosdahl** [comment]
> During development in a sufficiently complex project with multiple include directories (`-I`), headers files may move from one directory to another or may be shadowed by additional files with the same name and `#include_next`. In such cases _direct mode_ often produces false hits and the new file is not picked up. Should direct mode support such use cases or is it common practice to configure `direct_mode = false`?

It's known and [documented](https://ccache.dev/manual/4.4.html#_caveats) that the direct mode doesn't pick up such headers. It would of course be highly preferable if this limitation didn't exist, but I don't know how to fix it in a good way. Previous discussion: https://www.mail-archive.com/ccache@lists.samba.org/msg00920.html

I have no idea what is common practice, but I would assume that it's not that common since this is a rare edge case.
